File: python/benchmarks/bench_hashmap.py
import sys
import os
import random
import time

_script_dir = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, _script_dir)
sys.path.insert(0, os.path.join(_script_dir, ".."))
from src.hashmap import HashMap
from bench_common import SCALES, NUM_RUNS, get_memory_mb, mean_std, write_csv, get_results_dir

LOW_ENTROPY_CAPACITY = 64  # low-entropy / near-collision: few buckets, many keys
LOAD_FACTOR_N = 100_000
LOAD_FACTORS = [0.25, 0.5, 0.75, 1.0]


def run_one_insert(N, capacity):
    keys = list(range(N))
    random.shuffle(keys)
    map1 = HashMap(capacity=max(16, capacity))
    for k in keys:
        map1.insert(k, k)


def run_one_get(N, capacity):
    keys = list(range(N))
    random.shuffle(keys)
    map1 = HashMap(capacity=max(16, capacity))
    for k in keys:
        map1.insert(k, k)
    for k in keys:
        _ = map1.get(k)


def measure_insert_times(N, capacity, warmup=True):
    if warmup:
        run_one_insert(N, capacity)
    times_ms = []
    for _ in range(NUM_RUNS):
        start = time.perf_counter()
        run_one_insert(N, capacity)
        times_ms.append((time.perf_counter() - start) * 1000)
    return times_ms


def measure_get_times(N, capacity, warmup=True):
    if warmup:
        run_one_get(N, capacity)
    times_ms = []
    for _ in range(NUM_RUNS):
        start = time.perf_counter()
        run_one_get(N, capacity)
        times_ms.append((time.perf_counter() - start) * 1000)
    return times_ms


def main():
    results_dir = get_results_dir()
    os.makedirs(results_dir, exist_ok=True)

    # ---- Main scenario: scaled, random keys, with std dev and memory ----
    rows_main = [["N", "insert_mean_ms", "insert_std_ms", "get_mean_ms", "get_std_ms", "memory_mb"]]
    for N in SCALES:
        capacity = max(16, N)
        mem_before = get_memory_mb()
        insert_times = measure_insert_times(N, capacity, warmup=True)
        get_times = measure_get_times(N, capacity, warmup=True)
        mem_after = get_memory_mb()
        memory_mb = max(0.0, mem_after - mem_before) if mem_after else 0.0

        i_mean, i_std = mean_std(insert_times)
        g_mean, g_std = mean_std(get_times)
        rows_main.append([N, f"{i_mean:.6f}", f"{i_std:.6f}", f"{g_mean:.6f}", f"{g_std:.6f}", f"{memory_mb:.4f}"])
        print(f"N={N}: Insert {i_mean:.6f} ± {i_std:.6f} ms, Get {g_mean:.6f} ± {g_std:.6f} ms, memory={memory_mb:.4f} MB")
    write_csv(rows_main, results_dir, "python_hashmap.csv")

    # ---- Low-entropy / near-collision: small fixed capacity ----
    rows_low_entropy = [["N", "insert_mean_ms", "insert_std_ms", "get_mean_ms", "get_std_ms"]]
    for N in SCALES:
        insert_times = measure_insert_times(N, LOW_ENTROPY_CAPACITY, warmup=True)
        get_times = measure_get_times(N, LOW_ENTROPY_CAPACITY, warmup=True)
        i_mean, i_std = mean_std(insert_times)
        g_mean, g_std = mean_std(get_times)
        rows_low_entropy.append([N, f"{i_mean:.6f}", f"{i_std:.6f}", f"{g_mean:.6f}", f"{g_std:.6f}"])
        print(f"Low-entropy N={N}: Insert {i_mean:.6f} ± {i_std:.6f} ms, Get {g_mean:.6f} ± {g_std:.6f} ms")
    write_csv(rows_low_entropy, results_dir, "python_hashmap_low_entropy.csv")

    # ---- Load factor sensitivity: fixed N, vary capacity ----
    rows_lf = [["load_factor", "insert_mean_ms", "insert_std_ms", "get_mean_ms", "get_std_ms"]]
    N = LOAD_FACTOR_N
    for lf in LOAD_FACTORS:
        capacity = max(16, int(N / lf))
        insert_times = measure_insert_times(N, capacity, warmup=True)
        get_times = measure_get_times(N, capacity, warmup=True)
        i_mean, i_std = mean_std(insert_times)
        g_mean, g_std = mean_std(get_times)
        rows_lf.append([lf, f"{i_mean:.6f}", f"{i_std:.6f}", f"{g_mean:.6f}", f"{g_std:.6f}"])
        print(f"LoadFactor={lf}: Insert {i_mean:.6f} ± {i_std:.6f} ms, Get {g_mean:.6f} ± {g_std:.6f} ms")
    write_csv(rows_lf, results_dir, "python_hashmap_loadfactor.csv")


if __name__ == "__main__":
    main()

File: cpp/benchmarks/bench_hashmap.cpp
#include "hashmap.hpp"
#include <algorithm>
#include <chrono>
#include <cmath>
#include <cstdlib>
#include <filesystem>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <numeric>
#include <random>
#include <string>
#include <vector>

#if defined(__linux__) || defined(__APPLE__)
#include <sys/resource.h>
#endif

template <typename T>
inline void do_not_optimize(const T& value) {
    asm volatile("" : : "g"(value) : "memory");
}

namespace fs = std::filesystem;

static constexpr int SCALES[] = {1000, 10'000, 100'000, 1'000'000};
static constexpr int NUM_RUNS = 5;
static constexpr size_t LOW_ENTROPY_CAPACITY = 64;  // low-entropy / near-collision: few buckets
static constexpr int LOAD_FACTOR_N = 100'000;
static constexpr double LOAD_FACTORS[] = {0.25, 0.5, 0.75, 1.0};

static double memory_mb() {
#if defined(__linux__) || defined(__APPLE__)
    struct rusage u;
    if (getrusage(RUSAGE_SELF, &u) == 0)
        return u.ru_maxrss / 1024.0;  // KB -> MB (Linux); macOS is bytes, adjust if needed
#endif
    return 0.0;
}

static void mean_std(const std::vector<double>& v, double& mean, double& stddev) {
    mean = 0;
    for (double x : v) mean += x;
    mean /= v.size();
    stddev = 0;
    if (v.size() >= 2) {
        for (double x : v) stddev += (x - mean) * (x - mean);
        stddev = std::sqrt(stddev / (v.size() - 1));
    }
}

int main() {
    const char* results_env = std::getenv("RESULTS_DIR");
    std::string out_dir = results_env ? results_env : "../results/raw";
    fs::create_directories(out_dir);

    std::random_device rd;
    std::mt19937 g(rd());

    // ---- Main scenario ----
    std::string csv_path = out_dir + "/cpp_hashmap.csv";
    std::ofstream file(csv_path);
    if (!file.is_open()) {
        std::cerr << "Failed to open " << csv_path << "\n";
        return 1;
    }
    file << "N,insert_mean_ms,insert_std_ms,get_mean_ms,get_std_ms,memory_mb\n";
    file << std::fixed << std::setprecision(6);

    for (int n : SCALES) {
        std::vector<int> keys(n);
        std::iota(keys.begin(), keys.end(), 0);
        std::shuffle(keys.begin(), keys.end(), g);

        {  // warm-up
            hashmap::HashMap map(static_cast<size_t>(std::max(16, n)));
            for (int k : keys) map.insert(k, k);
            for (int k : keys) do_not_optimize(map.get(k));
        }

        std::vector<double> insert_ms(NUM_RUNS), get_ms(NUM_RUNS);
        for (int run = 0; run < NUM_RUNS; run++) {
            std::shuffle(keys.begin(), keys.end(), g);
            hashmap::HashMap map(static_cast<size_t>(std::max(16, n)));
            auto start = std::chrono::high_resolution_clock::now();
            for (int k : keys) map.insert(k, k);
            insert_ms[run] = std::chrono::duration<double, std::milli>(std::chrono::high_resolution_clock::now() - start).count();
            start = std::chrono::high_resolution_clock::now();
            for (int k : keys) do_not_optimize(map.get(k));
            get_ms[run] = std::chrono::duration<double, std::milli>(std::chrono::high_resolution_clock::now() - start).count();
        }
        double i_mean, i_std, g_mean, g_std;
        mean_std(insert_ms, i_mean, i_std);
        mean_std(get_ms, g_mean, g_std);
        double mem = memory_mb();
        file << n << "," << i_mean << "," << i_std << "," << g_mean << "," << g_std << "," << std::setprecision(4) << mem << "\n";
        file << std::setprecision(6);
        std::cout << "N=" << n << ": Insert " << i_mean << " ± " << i_std << " ms, Get " << g_mean << " ± " << g_std << " ms, memory=" << mem << " MB\n";
    }
    std::cout << "Wrote " << csv_path << "\n";
    file.close();

    // ---- Low-entropy / near-collision ----
    csv_path = out_dir + "/cpp_hashmap_low_entropy.csv";
    file.open(csv_path);
    file << "N,insert_mean_ms,insert_std_ms,get_mean_ms,get_std_ms\n";
    file << std::fixed << std::setprecision(6);
    for (int n : SCALES) {
        std::vector<int> keys(n);
        std::iota(keys.begin(), keys.end(), 0);
        std::shuffle(keys.begin(), keys.end(), g);
        { hashmap::HashMap map(LOW_ENTROPY_CAPACITY); for (int k : keys) map.insert(k, k); for (int k : keys) do_not_optimize(map.get(k)); }
        std::vector<double> insert_ms(NUM_RUNS), get_ms(NUM_RUNS);
        for (int run = 0; run < NUM_RUNS; run++) {
            std::shuffle(keys.begin(), keys.end(), g);
            hashmap::HashMap map(LOW_ENTROPY_CAPACITY);
            auto start = std::chrono::high_resolution_clock::now();
            for (int k : keys) map.insert(k, k);
            insert_ms[run] = std::chrono::duration<double, std::milli>(std::chrono::high_resolution_clock::now() - start).count();
            start = std::chrono::high_resolution_clock::now();
            for (int k : keys) do_not_optimize(map.get(k));
            get_ms[run] = std::chrono::duration<double, std::milli>(std::chrono::high_resolution_clock::now() - start).count();
        }
        double i_mean, i_std, g_mean, g_std;
        mean_std(insert_ms, i_mean, i_std);
        mean_std(get_ms, g_mean, g_std);
        file << n << "," << i_mean << "," << i_std << "," << g_mean << "," << g_std << "\n";
        std::cout << "Low-entropy N=" << n << ": Insert " << i_mean << " ± " << i_std << " ms, Get " << g_mean << " ± " << g_std << " ms\n";
    }
    std::cout << "Wrote " << csv_path << "\n";
    file.close();

    // ---- Load factor sensitivity ----
    csv_path = out_dir + "/cpp_hashmap_loadfactor.csv";
    file.open(csv_path);
    file << "load_factor,insert_mean_ms,insert_std_ms,get_mean_ms,get_std_ms\n";
    file << std::fixed << std::setprecision(6);
    int n = LOAD_FACTOR_N;
    std::vector<int> keys(n);
    std::iota(keys.begin(), keys.end(), 0);
    for (double lf : LOAD_FACTORS) {
        std::shuffle(keys.begin(), keys.end(), g);
        size_t cap = std::max(size_t(16), size_t(n / lf));
        { hashmap::HashMap map(cap); for (int k : keys) map.insert(k, k); for (int k : keys) do_not_optimize(map.get(k)); }
        std::vector<double> insert_ms(NUM_RUNS), get_ms(NUM_RUNS);
        for (int run = 0; run < NUM_RUNS; run++) {
            std::shuffle(keys.begin(), keys.end(), g);
            hashmap::HashMap map(cap);
            auto start = std::chrono::high_resolution_clock::now();
            for (int k : keys) map.insert(k, k);
            insert_ms[run] = std::chrono::duration<double, std::milli>(std::chrono::high_resolution_clock::now() - start).count();
            start = std::chrono::high_resolution_clock::now();
            for (int k : keys) do_not_optimize(map.get(k));
            get_ms[run] = std::chrono::duration<double, std::milli>(std::chrono::high_resolution_clock::now() - start).count();
        }
        double i_mean, i_std, g_mean, g_std;
        mean_std(insert_ms, i_mean, i_std);
        mean_std(get_ms, g_mean, g_std);
        file << lf << "," << i_mean << "," << i_std << "," << g_mean << "," << g_std << "\n";
        std::cout << "LoadFactor=" << lf << ": Insert " << i_mean << " ± " << i_std << " ms, Get " << g_mean << " ± " << g_std << " ms\n";
    }
    std::cout << "Wrote " << csv_path << "\n";
    return 0;
}
